### DISCUSSION TITLE
magit & consult don't play well when selecting a branch to get logs from

### DISCUSSION THREAD

**Discussion by sergiodj** [magit/magit#5173] [Support] [answered]
Hi there,

I'm filing this separate bug here (instead of against consult) because #4437 seems to be related, and it ended up being an issue with magit.

What I'm experiencing is actually a pretty similar problem, but in a different scenario and with a slightly different error.  First of all, I'm using a pristine Emacs instance launched with `emacs -Q` and using an empty `.emacs.d` directory where I'm instructing Emacs to install only magit and consult.  Then:

* I open a git repository with magit.
* I select the option `l o` (Log -> other).
* I type `TAB` to open up the completion menu.
* I type `M-<down>` to start cycling through the options.

When I select the first option, it correctly appears in the minibuffer.  But when I continue typing `M-<down>` to select other options, I see them being prepended to the existing option present in the minibuffer.  This makes things very confusing, because I can't just select an option and hit `RET` to open the log buffer; instead, I have to manually adjust the minibuffer and remove all the cruft that's been inserted there.

As I said, I'm not entirely sure whether this is a bug in magit or consult, and I'll be happy to file the same bug somewhere else if applicable.

Thanks in advance.

**Comment by sergiodj** [comment]
Another interesting data point.  To the extent of what I was able to test, this only seems to happen when selecting a branch in the `l o` option.  If I, for example, go to the log buffer and type `A A` to select a commit to cherry-pick, and then hit `TAB`, I can navigate through the options using `M-<down>` and `M-<up>` normally.

**Comment by tarsius** [comment]
I cannot reproduce this.

I have installed consult too and that doesn't appear to be enough for it to be used with magit. I have grepped consult for "magit" and it does not appear once.

Do you use any additional configuration that you did not mention? Such as the `(advice-add #'completing-read-multiple :override #'consult-completing-read-multiple)` mentioned in #4437.

As noted at https://github.com/magit/magit/issues/4437#issuecomment-1119843297 that was removed from consult, and in response I removed the respective kludge from magit. Are you maybe using an old version of consult?

**Reply by sergiodj** [reply]
Thanks for the reply.

No, I don't use any advices.  Here's the configuration I'm using for magit and consult:

```elisp
(require 'package)
(add-to-list 'package-archives
	     '("melpa" . "https://melpa.org/packages/"))
(add-to-list 'package-archives
	     '("mepa-stable" . "https://stable.melpa.org/packages/"))

(use-package magit
  :ensure t
  :bind
  ("C-x g" . #'magit-status)
  :config
  (setq magit-diff-refine-hunk 'all))

(use-package consult
  :ensure t
  ;; Replace bindings. Lazily loaded due by `use-package'.
  :bind (;; C-c bindings in `mode-specific-map'
         ("C-c M-x" . consult-mode-command)
         ("C-c h" . consult-history)
         ("C-c k" . consult-kmacro)
         ("C-c m" . consult-man)
         ("C-c i" . consult-info)
         ([remap Info-search] . consult-info)
         ;; C-x bindings in `ctl-x-map'
         ("C-x M-:" . consult-complex-command)     ;; orig. repeat-complex-command
         ("C-x b" . consult-buffer)                ;; orig. switch-to-buffer
         ("C-x 4 b" . consult-buffer-other-window) ;; orig. switch-to-buffer-other-window
         ("C-x 5 b" . consult-buffer-other-frame)  ;; orig. switch-to-buffer-other-frame
         ("C-x t b" . consult-buffer-other-tab)    ;; orig. switch-to-buffer-other-tab
         ("C-x r b" . consult-bookmark)            ;; orig. bookmark-jump
         ("C-x p b" . consult-project-buffer)      ;; orig. project-switch-to-buffer
         ;; Custom M-# bindings for fast register access
         ("M-#" . consult-register-load)
         ("M-'" . consult-register-store)          ;; orig. abbrev-prefix-mark (unrelated)
         ("C-M-#" . consult-register)
         ;; Other custom bindings
;	 ("M-x" . consult-mode-command)
         ;("C-y" . consult-yank-from-kill-ring)                ;; orig. yank-pop
	 ("M-y" . consult-yank-replace)
         ;; M-g bindings in `goto-map'
	 ("C-s" . consult-line)
;         ("M-g e" . consult-compile-error)
;         ("M-g f" . consult-flymake)               ;; Alternative: consult-flycheck
         ("M-g g" . consult-goto-line)             ;; orig. goto-line
         ("M-g M-g" . consult-goto-line)           ;; orig. goto-line
         ("M-g o" . consult-outline)               ;; Alternative: consult-org-heading
         ("M-g m" . consult-mark)
         ("M-g k" . consult-global-mark)
         ("M-g i" . consult-imenu)
         ("M-g I" . consult-imenu-multi)
         ;; M-s bindings in `search-map'
         ("M-s d" . consult-find)                  ;; Alternative: consult-fd
         ("M-s c" . consult-locate)
         ("M-s g" . consult-grep)
         ("M-s G" . consult-git-grep)
         ("M-s r" . consult-ripgrep)
         ("M-s l" . consult-line)
         ("M-s L" . consult-line-multi)
         ("M-s k" . consult-keep-lines)
         ("M-s u" . consult-focus-lines)
         ;; Isearch integration
;         ("M-s e" . consult-isearch-history)
         ;; :map isearch-mode-map
         ;; ("M-e" . consult-isearch-history)         ;; orig. isearch-edit-string
         ;; ("M-s e" . consult-isearch-history)       ;; orig. isearch-edit-string
         ;; ("M-s l" . consult-line)                  ;; needed by consult-line to detect isearch
         ;; ("M-s L" . consult-line-multi)            ;; needed by consult-line to detect isearch
         ;; Minibuffer history
         :map minibuffer-local-map
         ("M-s" . consult-history)                 ;; orig. next-matching-history-element
         ("M-r" . consult-history))                ;; orig. previous-matching-history-element

  ;; Enable automatic preview at point in the *Completions* buffer. This is
  ;; relevant when you use the default completion UI.
  :hook (completion-list-mode . consult-preview-at-point-mode)

  ;; The :init configuration is always executed (Not lazy)
  :init

  ;; Use Consult to select xref locations with preview
  (setq xref-show-xrefs-function #'consult-xref
        xref-show-definitions-function #'consult-xref)

  ;; Configure other variables and modes in the :config section,
  ;; after lazily loading the package.
  :config

  ;; Optionally configure preview. The default value
  ;; is 'any, such that any key triggers the preview.
  ;; (setq consult-preview-key 'any)
  ;; (setq consult-preview-key "M-.")
  ;; (setq consult-preview-key '("S-<down>" "S-<up>"))
  ;; For some commands and buffer sources it is useful to configure the
  ;; :preview-key on a per-command basis using the `consult-customize' macro.
  (consult-customize
   consult-theme :preview-key '(:debounce 0.2 any)
   consult-ripgrep consult-git-grep consult-grep
   consult-bookmark consult-recent-file consult-xref
   consult--source-bookmark consult--source-file-register
   consult--source-recent-file consult--source-project-recent-file
   ;; :preview-key "M-."
   :preview-key '(:debounce 0.4 any))

  ;; Optionally configure the narrowing key.
  ;; Both < and C-+ work reasonably well.
  (setq consult-narrow-key "<") ;; "C-+"
)
```

The consult configuration is basically the one suggested by upstream.

I'm using `GNU Emacs 30.0.60 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.42, cairo version 1.18.0) of 2024-06-25`, btw.

**Reply by sergiodj** [reply, marked as answer]
Hm, I've just tested this on Emacs 29.3 and I can't reproduce the problem.  Since I'm building my own Emacs, it seems there's something that happened on the `master` branch that affected consult & magit.  I'm now building a newer version of Emacs here to see if the problem still happens, but it doesn't seem to be an issue with magit.  Sorry about the noise.
